src/Tickets/QR/Controller.php

```diff
@@ -0,0 +1,172 @@
+<?php
+
+namespace TEC\Tickets\QR;
+
+use \TEC\Common\Contracts\Provider\Controller as Controller_Contract;
+
+/**
+ * Class Controller.
+ *
+ * @since   TBD
+ *
+ * @package TEC\Tickets\QR
+ */
+class Controller extends Controller_Contract {
+	/**
+	 * Determines if this controller will register.
+	 * This is present due to how UOPZ works, it will fail if method belongs to the parent/abstract class.
+	 *
+	 * @since TBD
+	 *
+	 * @return bool
+	 */
+	public function is_active(): bool {
+		return true;
+	}
+
+	/**
+	 * Boot the Controller.
+	 *
+	 * This function is used to instantiate the singleton classes and register any other providers.
+	 *
+	 * @since   TBD
+	 */
+	public function boot() {
+		$this->container->bind( Facade::class, [ $this, 'bind_facade_or_error' ] );
+	}
+
+	/**
+	 * Register the controller.
+	 *
+	 * @since   TBD
+	 */
+	public function do_register(): void {
+		$this->boot();
+
+		$this->add_actions();
+		$this->add_filters();
+	}
+
+	/**
+	 * Unregister the controller.
+	 *
+	 * @since   TBD
+	 *
+	 * @return void
+	 */
+	public function unregister(): void {
+		$this->remove_actions();
+		$this->remove_filters();
+	}
+
+	/**
+	 * Add the action hooks.
+	 *
+	 * @since   TBD
+	 *
+	 * @return void
+	 */
+	protected function add_actions(): void {
+
+	}
+
+	/**
+	 * Add the action hooks.
+	 *
+	 * @since   TBD
+	 *
+	 * @return void
+	 */
+	protected function add_filters(): void {
+
+	}
+
+	/**
+	 * Removes actions.
+	 *
+	 * @since   TBD
+	 *
+	 * @return void
+	 */
+	protected function remove_actions(): void {
+
+	}
+
+	/**
+	 * Removes filters.
+	 *
+	 * @since   TBD
+	 *
+	 * @return void
+	 */
+	protected function remove_filters(): void {
+
+	}
+
+	/**
+	 * Binds the facade or throws an error.
+	 *
+	 * @since TBD
+	 *
+	 * @return \WP_Error|Facade
+	 */
+	public function bind_facade_or_error() {
+		if ( ! $this->can_use() ) {
+			return new \WP_Error(
+				'tec_tickets_qr_code_cannot_use',
+				__( 'The QR code cannot be used, please contact your host and ask for `gzip` and `gd` support.', 'event-tickets' )
+			);
+		}
+
+		// Load the library if it's not loaded already.
+		$this->load_library();
+
+		return new Facade;
+	}
+
+	/**
+	 * Determines if the QR code library is loaded.
+	 *
+	 * @since TBD
+	 *
+	 * @return void
+	 */
+	protected function has_library_loaded(): bool {
+		return class_exists( 'QRencode', false ) || class_exists( 'QRcode', false );
```

Comment:
If I understand this correctly, the idea is to avoid auto-loading the library from **another** source that might provide it (e.g. another plugin); but this will not avoid a conflict with that library.
If it's already loaded, it might not be the version required by ET.
If it's not loaded, and is not loaded at this stage since autoloading is omitted, then loading the ET version of the library would not avoid a possible version conflict.
Since the library is shipped with the plugin anyway, why not vendor-prefix the QR library and just always load the vendor-prefixed version?

Reply:
Agreed. We definitely need to do vendor prefixing